tests/test_atproto_client/models/tests/test_string_formats.py

```diff
@@ -0,0 +1,198 @@
+        (string_formats.Tid, 'tid', 'must be exactly 13 lowercase letters/numbers'),
+        (string_formats.RecordKey, 'record_key', 'must contain only alphanumeric'),
+        (string_formats.Uri, 'uri', 'must be a valid URI with scheme and authority/path'),
+        (string_formats.Language, 'language', 'must be ISO language code'),
+    ],
+)
+def test_string_format_validation(
+    validator_type: type, field_name: str, expected_error: str, valid_data: dict, invalid_data: dict
+) -> None:
+    """Test validation for each string format type."""
+    SomeTypeAdapter = TypeAdapter(validator_type)
+
+    # Test that validation is skipped by default
+    assert SomeTypeAdapter.validate_python(invalid_data[field_name]) == invalid_data[field_name]
+
+    # Test that valid data passes strict validation
+    validated_value = SomeTypeAdapter.validate_python(valid_data[field_name], context={_OPT_IN_KEY: True})
+    assert validated_value == valid_data[field_name]
+
+    # Test that invalid data fails strict validation
+    try:
+        SomeTypeAdapter.validate_python(invalid_data[field_name], context={_OPT_IN_KEY: True})
+        pytest.fail(f'{validator_type.__name__} validation should have failed')
+    except ValidationError as e:
+        error = e.errors()[0]
+        assert expected_error in error['msg']
+
+
+@pytest.mark.parametrize(
+    'valid_value',
+    [
+        'test.bsky.social',
+        'did:plc:1234abcd',
+        'app.bsky.feed.post',
+        'at://user.bsky.social/posts/123',
+        'bafyreidfayvfuwqa2beehqn7axeeeaej5aqvaowxgwcdt2rw',
+        '2023-01-01T12:00:00Z',
+        '3jqvenncqkgbm',
+        'valid-key',
+        'https://example.com',
+        'en-US',
+    ],
+)
+def test_generic_string_format_validation(valid_value: str) -> None:
+    """Test that ATProtoString accepts each valid string format."""
+
+    validated = TypeAdapter(string_formats.ATProtoString).validate_python(valid_value, context={_OPT_IN_KEY: True})
+    assert validated == valid_value
+
+
+def test_get_or_create_with_strict_validation(valid_data: dict, invalid_data: dict) -> None:
+    """Test that get_or_create uses strict validation."""
+
+    class TestModel(BaseModel):
+        handle: string_formats.Handle
+        did: string_formats.Did
+
+    # Test valid data passes
+    instance = get_or_create(
+        {'handle': valid_data['handle'], 'did': valid_data['did']}, TestModel, strict_string_format=True
+    )
+    assert isinstance(instance, TestModel)
+    assert instance.handle == valid_data['handle']
+    assert instance.did == valid_data['did']
+
+    # Test invalid handle fails
+    try:
+        get_or_create(
+            {'handle': invalid_data['handle'], 'did': valid_data['did']}, TestModel, strict_string_format=True
+        )
+        pytest.fail('Handle validation should have failed')
+    except ModelError as e:
+        assert 'must be a domain name' in str(e)
```

Comment:
pytest has syntax sugar to not deal with exceptions in try-except blocks. check this one:

https://github.com/MarshalX/atproto/blob/8aafc9a4019b0a6cd68da73a27e5f4355225d102/tests/test_atproto_client/models/tests/test_changed_lexicon_compatability.py#L71-L80

it will automatically handle if exception is not occurred and will properly fail the test